### PR TITLE
docs: clarification for Unix.open_process_args_in

### PR DESCRIPTION
**Pull request by rbjorklin**
I tripped myself up on this [as can be seen here](https://discuss.ocaml.org/t/is-unix-open-process-args-in-broken/6865/3). Hoping this small change will help others in the future.

I used the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) style, I hope that's okay.

### CODE REVIEW

**Comment by gasche** [comment]
Searching the web suggests that you are far from the only one to make this mistake. See for example [this piece of Reason code](https://github.com/onivim/oni2/blob/222a98aee3a95b683d5d8448b21d777732ef0418/src/Exthost/Utility.re)

```reason
  let ic =
    Sys.win32
      // HACK: Not sure why this command doesn't work on Linux / macOS, and vice versa...
      ? Unix.open_process_args_in(
          "node",
          [|"node", "-e", "console.log(process.execPath)"|],
        )
      : Unix.open_process_in("node -e 'console.log(process.execPath)'");
```

It looks like Windows resolves the executable name passed to open_process_args*, adding to the confusion.

Could you maybe consider adding an example use in the documentation, which would make the recommendations visually obvious? Code blocks in ocamldoc use the `{[...]}` markup.

.

I wonder if we should consider changing `open_process_args*` to lookup the executable in the PATH (there is logic in the Unix module to implement this easily). My understanding of the function was that it was introduced to avoid the complexity of *argument* parsing by the shell, but there is not much in the original PR #1792 about resolution of the executable itself.

An alternative since OCaml 4.10 is to use `Sys.command (Filename.quote_command cmd args)`. If I understand correctly, this lets us pass arguments as an unescaped list, but the `cmd` command will still be resolved through the shell.

**Comment by gasche** [comment]
(cc @xavierleroy @nojb)

**Comment by nojb** [comment]
> I wonder if we should consider changing `open_process_args*` to lookup the executable in the PATH (there is logic in the Unix module to implement this easily). My understanding of the function was that it was introduced to avoid the complexity of _argument_ parsing by the shell, but there is not much in the original PR #1792 about resolution of the executable itself.

Indeed, to summarize my recollection:

- `open_process{,_in,_out,_full}`: builds a shell command line and executes through shell; as a side-effect, the program is looked up in `PATH`
- `open_process_args{,_in,_out,_full}`: executes the program with the given arguments "verbatim". This uses `execv` on Unix so the program is not looked up in the `PATH`. However, on Windows (which requires an ad-hoc implementation), the program *is* looked up in the `PATH`, see https://github.com/ocaml/ocaml/blob/845ac6a50d54d929bfcfc6df10215e04d449cdac/otherlibs/win32unix/createprocess.c#L98

So indeed it looks like there is a mismatch between Unix/Windows regarding the *args* functions. This may have been overlooked when the *args* functions where introduced because up until that point, only `/bin/sh` was used as "program" argument to the `exec*` function, which didn't need to be looked up in `PATH`.

So to summarize, looks like replacing `execv` by `execvp` in the Unix implementation of the *args* functions shwould be a plausible fix for the present issue.

**Review by xavierleroy — COMMENTED** [review]
The proposed documentation is correct.  However, the convention on argument arrays (that the first element matches the command name and the subsequent elements are the arguments proper) is consistently used in the Unix library, e.g. in `execv`, `execve`, `execvp`, `execvpe`, `create_process`, `create_process_env`, `open_process_args_in`, `open_process_args_out`, `open_process_args`, and `open_process_args_full`.  So, documenting it for `open_process_args_in` and not for any other function is problematic.

**Comment by xavierleroy** [comment]
> So to summarize, looks like replacing execv by execvp in the Unix implementation of the args functions shwould be a plausible fix for the present issue.

It is a plausible change, making these functions more usable, with a small backward incompatibility.

However it is not a "fix for the present issue", which also includes a misunderstanding of what an "argument array" is in the Unix world.  This is a bit of information that should be put somewhere but not repeated 10 times for the 10 Unix functions that are affected.

**Comment by frou** [comment]
I created a [separate issue](https://github.com/ocaml/ocaml/issues/10083) for the `open_process_args*` and PATH aspect. @gasche @nojb

**Comment by rbjorklin** [comment]
Superseded by #10084.
